arangod/Aql/QueryCache.cpp

```diff
@@ -84,27 +191,64 @@ std::shared_ptr<QueryCacheResultEntry> QueryCacheDatabaseEntry::lookup(
   }
 
   // found some result in cache
+  auto entry = (*it).second.get();
 
-  if (queryString.size() != (*it).second->_queryString.size() ||
-      memcmp(queryString.data(), (*it).second->_queryString.c_str(), queryString.size()) != 0) {
+  if (queryString.size() != entry->_queryString.size() ||
+      memcmp(queryString.data(), entry->_queryString.data(), queryString.size()) != 0) {
     // found something, but obviously the result of a different query with the
     // same hash
     return nullptr;
   }
 
+  // compare bind variables
+  VPackSlice entryBindVars = VPackSlice::emptyObjectSlice();
+  if (entry->_bindVars != nullptr) {
+    entryBindVars = entry->_bindVars->slice();
+  }
+  VPackValueLength entryLength = entryBindVars.length();
+  
+  VPackSlice lookupBindVars = VPackSlice::emptyObjectSlice();
+  if (bindVars != nullptr) {
+    lookupBindVars = bindVars->slice();
+  }
+  VPackValueLength lookupLength = lookupBindVars.length();
+
+  if (entryLength > 0 || lookupLength > 0) {
+    if (entryLength != lookupLength) {
+      // different number of bind variables
+      return nullptr;
+    }
+
+    if (basics::VelocyPackHelper::compare(entryBindVars, lookupBindVars, false) != 0) {
+      // different bind variables
+      return nullptr;
+    }
+  }
+
+  // all equal -> hit!
+  entry->increaseHits();
+
   // found an entry
   return (*it).second;
 }
 
 /// @brief store a query result in the database-specific cache
-void QueryCacheDatabaseEntry::store(uint64_t hash,
-                                    std::shared_ptr<QueryCacheResultEntry> entry) {
+void QueryCacheDatabaseEntry::store(std::shared_ptr<QueryCacheResultEntry>&& entry,
```

Comment:
I think it would be enough to std::move() you don't need to force it (unless thats intended)

Reply:
You are right. But we only have one caller which is happy to give up the ownership in all cases, so I think this is just fine.